lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,354 @@
+
+      self.homepage_url = 'https://shift4.com'
+      self.display_name = 'Shift4'
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        'incorrect_number' => STANDARD_ERROR_CODE[:incorrect_number],
+        'invalid_number' => STANDARD_ERROR_CODE[:invalid_number],
+        'invalid_expiry_month' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'invalid_expiry_year' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'invalid_cvc' => STANDARD_ERROR_CODE[:invalid_cvc],
+        'expired_card' => STANDARD_ERROR_CODE[:expired_card],
+        'insufficient_funds' => STANDARD_ERROR_CODE[:card_declined],
+        'incorrect_cvc' => STANDARD_ERROR_CODE[:incorrect_cvc],
+        'incorrect_zip' => STANDARD_ERROR_CODE[:incorrect_zip],
+        'card_declined' => STANDARD_ERROR_CODE[:card_declined],
+        'processing_error' => STANDARD_ERROR_CODE[:processing_error],
+        'lost_or_stolen' => STANDARD_ERROR_CODE[:card_declined],
+        'suspected_fraud' => STANDARD_ERROR_CODE[:card_declined],
+        'expired_token' => STANDARD_ERROR_CODE[:card_declined]
+      }
+
+      def initialize(options = {})
+        requires!(options, :client_guid, :auth_token)
+        super
+      end
+
+      def access_token(options = {})
+        post = {}
+        add_credentials(post, options)
+        commit('accesstoken', post, request_headers(options))
+      end
+
+      def purchase(money, authorization, options = {})
+        post = {}
+        add_datetime(post, options)
+        api_options(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, options)
+        add_transaction(post, options)
+        add_card(post, authorization, options)
+
+        commit('sale', post, request_headers(options))
+      end
+
+      def authorize(money, card, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, options)
+        add_transaction(post, options)
+        add_card(post, card, options)
+        add_customer(post, options)
+
+        commit('authorization', post, request_headers(options))
+      end
+
+      def capture(money, authorization, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, options)
+        add_transaction(post, options)
+        add_card(post, authorization, options)
+
+        commit('capture', post, request_headers(options))
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        add_datetime(post, options)
+        api_options(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, options)
+        add_transaction(post, options)
+        add_device(post, options)
+        add_p2pe(post, options)
+        add_emv(post, options)
+        add_customer(post, options)
+        add_card(post, authorization, options)
+
+        commit('refund', post, request_headers(options))
```

Comment:
Small nit, but usually we pass `options` to the `commit` method and then call the `headers` method within the `commit` function. This is useful in cases where `options` is needed in that method.